pkg/controllers/hostedcluster/controller/controller.go

```diff
@@ -0,0 +1,215 @@
+	// hostedClusterImportFinalizer is the finalizer added to hostedcluster for importing as managed cluster
+	hostedClusterImportFinalizer = "cluster.open-cluster-management.io/import-managedcluster-finalizer"
+)
+
+// HostedClusterReconciler reconciles a HostedCluster object
+type HostedClusterReconciler struct {
+	client.Client
+	// Scheme *runtime.Scheme
+	ControlplaneKubeClient     kubernetes.Interface
+	ControlplaneOperatorClient operatorclient.Interface
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(1).Info("Reconciling hostecluster", "namespace", req.NamespacedName.Namespace, "name", req.NamespacedName.Name)
+
+	hostedCluster := &hyperv1beta1.HostedCluster{}
+	if err := r.Get(ctx, req.NamespacedName, hostedCluster); err != nil {
+		log.Error(err, "unable to fetch hostedcluster")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// hostedcluster is terminating, detach the managed cluster and return
+	if !hostedCluster.GetDeletionTimestamp().IsZero() {
+		if containStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer) {
+			if err := r.pruneManagedClusterResources(ctx, log, hostedCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			hostedCluster.SetFinalizers(removeStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer))
+			if err := r.Client.Update(ctx, hostedCluster, &client.UpdateOptions{}); err != nil {
+				if errors.IsConflict(err) {
+					log.Info("conflict when removing finalizer from hostedcluster instance", "hostedcluster", req.NamespacedName)
+					return ctrl.Result{Requeue: true}, nil
+				} else if err != nil {
+					log.Error(err, "unable to remove finalizer to hostedcluster instance", "hostedcluster", req.NamespacedName)
+					return ctrl.Result{}, err
+				}
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	if !containStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer) {
+		hostedCluster.SetFinalizers(append(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer))
+		if err := r.Client.Update(ctx, hostedCluster, &client.UpdateOptions{}); err != nil {
+			if errors.IsConflict(err) {
+				log.Info("conflict when adding finalizer to hostedcluster instance", "hostedcluster", req.NamespacedName)
+				return ctrl.Result{Requeue: true}, nil
+			} else if err != nil {
+				log.Error(err, "unable to add finalizer to hostedcluster instance", "hostedcluster", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	var hostedClusterKubeconfigSecret corev1.Secret
```

Comment:
can we get the klusterlet firstly? if we found it, we can stop the reconcile to reduce the unnecessary operations